### PR TITLE
[LangRef] Clarify semantics of masked vector load/store

### PR DESCRIPTION
**Pull request by RalfJung**
Basically, these operations are equivalent to a loop that iterates all elements and then does a `getelementptr` (without `inbounds`!) plus `load`/`store` only for the masked-on elements.

### CODE REVIEW

File: llvm/docs/LangRef.rst
Comment (RalfJung) [inline llvm/docs/LangRef.rst:23755]: Is "masked-on" the opposite of "masked-off"? Or is there some other term I could use?

**Comment by nikic** [comment]
I would rephrase this in terms of something like this:

> However, these intrinsics behave as-is the masked off lanes are not accessed.

Which should tell use everything necessary about their semantics. Then can continue to clarify that this means no exceptions / data races / etc.

**Comment by RalfJung** [comment]
That doesn't quite say everything -- there's the question of whether this [Rust PR](https://github.com/rust-lang/rust/pull/119203) should say `offset` (aka `getelementptr inbounds`) or `offset_wrapping` (aka `getelementptr`) when describing how the pointers to the individual elements being loaded are computed.

**Comment by programmerjake** [comment]
> That doesn't quite say everything -- there's the question of whether this [Rust PR](https://github.com/rust-lang/rust/pull/119203) should say `offset` (aka `getelementptr inbounds`)

there is the additional caveat that LLVM is allowed to create a `poison` value without UB (which is what happens with `getelementptr inbounds` with out-of-bounds indexes), but Rust defines out-of-bounds `offset` to be immediate UB, instead of deferred to the load/store.

a major difference between the two choices is that doing a masked load on a pointer before the beginning of it's allocation is disallowed with `inbounds`, but allowed without `inbounds` as long as vector elements are masked-off until the offset is big enough to be in the allocation's bounds.

**Comment by RalfJung** [comment]
> a major difference between the two choices is that doing a masked load on a pointer before the beginning of it's allocation is disallowed with inbounds, but allowed without inbounds as long as vector elements are masked-off until the offset is big enough to be in the allocation's bounds.

Yes that is indeed the key point: if the first half of the vector is masked-off, and that first half is actually out-of-bounds, then the pointer itself is conceptually out-of-bounds and "computing the pointer to the actually loaded element" would be a non-inbounds pointer computation. I expect this usecase to be allowed, which is why I added

> Only the masked-on lanes of the vector need to be inbounds of an allocation (but all these lanes need to be inbounds of the same allocation).

**Comment by RalfJung** [comment]
@nikic I have updated the wording to

> The result of this operation is equivalent to a regular vector load instruction followed by a 'select' between the loaded and the passthru values, predicated on the same mask, except that the masked-off lanes are not accessed.

Followed by clarification regarding exceptions, noalias, and data races. Does that work for you?

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:24425]:
```suggestion
Masked-off lanes are also not considered accessed for the purpose of data races or ``noalias`` constraints.
```

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:24470]:
```suggestion
Masked-off lanes are also not considered accessed for the purpose of data races or ``noalias`` constraints.
```

**Review by nikic — APPROVED** [review]
LGTM, but a second opinion wouldn't hurt.

**Comment by RalfJung** [comment]
@nikic any recommendation for how one could get a second opinion for this PR? I don't know how to navigate the LLVM reviews process to move this PR forwards...

**Comment by RalfJung** [comment]
@nunoplopes any chance you could take a look at this? :)

**Comment by appujee** [comment]
cc: @fhahn  and @alexey-bataev  who are code owners of Autovectorizer

**Comment by alexey-bataev** [comment]
No objections from my side

**Comment by RalfJung** [comment]
Thanks! Could someone merge this please then? :)

**Comment by nikic** [comment]
> Thanks! Could someone merge this please then? :)

Sure, but you need to update the PR description first, which becomes the commit message.

**Comment by RalfJung** [comment]
@nikic I updated the description, does that work?

**Comment by nunoplopes** [comment]
LGTM

**Pull request merged**
